vc/dasharo/options.h: Expose descriptor writeability flag as EFI var

mkopec opened the pull request:
Also see https://github.com/Dasharo/edk2/pull/274

mkopec committed:
vc/dasharo/options.h: Expose descriptor writeability flag as EFI var

Tested by booting into Ubuntu 24.04 on a NovaCustom V540TU and checking
the state of the DescriptorWriteable EFI variable via efivarfs.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: Ife03e252f41a58af269f44c8e99e9c627c51c79d
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

miczyg1 commented on src/vendorcode/dasharo/options.c:
```diff
@@ -540,3 +543,68 @@ bool get_ibecc_option(bool ibecc_default)
+
+	printk(BIOS_DEBUG, "Descriptor is %swriteable\n", descriptor_writeable ? "" : "not ");
+
+	efi_fv_set_option(&smmstore_rdev,
```
I recall writing a variable was not yet supproted in the initial patchset of SMMSTORE EFI variables. I think you may have some problems here I, at least, had a lot if variables were READ multiple times in one stage, because rdev pointer was being overwritten and ended up at the end of storage when attempted to read variable 2nd time or more.

SergiiDmytruk replied:
We have your fix in https://github.com/Dasharo/coreboot/commit/788c03436b15, but I remember using it and still having trouble when trying to write XMP profile to an EFI variable.  It's possible that this write won't be affected by such issues.

By the way, coreboot never creates variable store, so nothing can be communicated via it on the first boot or if the store needs recovery.  I guess IFD's lock state is not exposed in any register?  We can't read IFD from EDK because SmmStore allows that only when capsules are present.

mkopec replied:
IFD lock can be read from the descriptor itself (which would require reading and parsing the descriptor in edk2) or from TGL onwards through HFSTS1 bit 4 (which isn't available while ME is disabled, unless https://github.com/Dasharo/coreboot/pull/453 is used)

mkopec replied:
@miczyg1 can this thread be resolved?

SergiiDmytruk replied:
We'll probably want to pass this data via cbmem in the future to have first boot behave the same as any other, but this should do for now.  If testing won't demonstrate any breakage, adding variables at the end of the ramstage may be working fine.  I was trying to do it in romstage.

BeataZdunczyk replied:
@SergiiDmytruk So creating a GH issue should be enough for this release, and with the next IBG implementation we can start addressing this issue?

SergiiDmytruk replied:
@BeataZdunczyk, yes, it's not a big deal, users probably won't even notice and the condition should be gone after a reboot anyway.  But it already confused me while testing and can also affect automated tests.

mkopec replied:
https://github.com/Dasharo/dasharo-issues/issues/1560

BeataZdunczyk replied:
>  But it already confused me while testing and can also affect automated tests.

Which tests? IBG: https://github.com/Dasharo/open-source-firmware-validation/blob/develop/dasharo-security/cbnt.robot?

miczyg1 replied:
HFSTS1 bit4 may not necessarily show the descriptor lock. E.g. if descriptor is locked, but ME is not, then this bit may be clear...

But I have good news for you. Check `BIOS_FDOC` and `BIOS_FDOD` in the SPI MMIO (PCH EDS). These registers can be used to read the descriptor without using the whole SPI chip reads, just plain and simple index/data pair. I believe it existed since ICH era...

@mkopec

SergiiDmytruk replied:
> Which tests? IBG: https://github.com/Dasharo/open-source-firmware-validation/blob/develop/dasharo-security/cbnt.robot?

@BeataZdunczyk, no, capsule tests and menu tests if a binary with locked IFD is used.  I don't think we do that at the moment.

mkopec committed:
vc/dasharo/options.c: Don't enable HMRFPO on capsule update

Don't issue HMRFPO if Capsule Update is enabled, as HMRFPO causes a
global reset which corrupts capsules.

Upstream-Status: Inappropriate [Dasharo downstream]
Change-Id: I2dc0384360c1028d547c572f47f8a767abc61fe2
Signed-off-by: Michał Kopeć <michal.kopec@3mdeb.com>

mkopec commented:
V540TU Capsule Update now passes. Needed to remove the HMRFPO disabling code (uncommented, to allow easy re-adding once needed again)

SergiiDmytruk commented on src/vendorcode/dasharo/options.c:
```diff
@@ -271,21 +274,21 @@ uint8_t cse_get_me_disable_mode(void)
-		else
-			disable_me_hmrfpo(&var);
-	}
+	//if (CONFIG(DASHARO_FIRMWARE_UPDATE_MODE) && fum) {
```
That's quite radical, could have just added `CONFIG(DRIVERS_EFI_UPDATE_CAPSULES)` and a comment that locked IFD implies no HAP during capsule updates and we must avoid cold reset on enabling HMRFPO.  The check could even be more fine grained and additionally check for IFD being locked.  I just worry of some unexpected change in behaviour during updates on other platforms.

mkopec replied:
Changed to disable HMRFPO if capsule update is enabled https://github.com/Dasharo/coreboot/pull/760/commits/c50cf495a488d2d1736a60cdd32198cc25579a7d
Decided not to include the IFD check as this issue affects all platforms using capsule update, not only the ones with a locked descriptor.

mkopec commented:
@SergiiDmytruk is there anything else missing here?

SergiiDmytruk reviewed (dismissed):
> @SergiiDmytruk is there anything else missing here?

No (other than CI), I was doing some tests with QEMU to verify UI updates and update blocking, seems to work as expected.

miczyg1 commented:
FYI: I already bumped edk2 rev here: https://github.com/Dasharo/coreboot/pull/762 (it is newer than the one updated here)
